components/builder-db/src/pool.rs

```diff
@@ -61,6 +77,57 @@ impl Pool {
             thread::sleep(Duration::from_millis(connection_retry_ms));
         }
     }
+
+    pub fn get_raw(&self)
+                   -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let conn = self.inner.get().map_err(Error::ConnectionTimeout)?;
+        Ok(conn)
+    }
+
+    pub fn get_shard
+        (&self,
+         shard_id: u32)
+         -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let conn = self.inner.get().map_err(Error::ConnectionTimeout)?;
+        debug!("Switching to shard {}", shard_id);
+
+        let schema_name = format!("shard_{}", shard_id);
+        let sql_search_path = format!("SET search_path TO {}", schema_name);
+        conn.execute(&sql_search_path, &[])
+            .map_err(Error::SchemaSwitch)?;
+        Ok(conn)
+    }
+
+    pub fn get<T: Routable>
+        (&self,
+         routable: &T)
+         -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let optional_shard_id = routable
+            .route_key()
+            .map(|k| k.hash(&mut FnvHasher::default()));
+
+        let shard_id = match optional_shard_id {
+            Some(id) => {
+                if id == 0 {
+                    let mut rng = rand::thread_rng();
+                    match rng.choose(&self.shards) {
+                        Some(shard) => *shard,
+                        None => 0,
+                    }
+                } else {
+                    (id % SHARD_COUNT as u64) as u32
+                }
+            }
+            None => {
+                let mut rng = rand::thread_rng();
+                match rng.choose(&self.shards) {
+                    Some(shard) => *shard,
+                    None => 0,
+                }
```

Comment:
Yeah this seemed a bit wierd to me as well.. if we pick a random id in this case, it seems we will get inconsistent behavior.  Not quite clear what scenario this is for.